server/src/tracing.ts

```diff
@@ -0,0 +1,202 @@
+  this: PatchedTransaction,
+  endTimestamp?: number
+) {
+  this._emitter?.emit("finish", this);
+  return _transactionFinish.call(this, endTimestamp);
+};
+
+// @ts-expect-error: _onFinish doesn't exist; we're adding it.
+Transaction.prototype._onFinish = function onFinish(
+  this: PatchedTransaction,
+  listener: (...args: any[]) => void
+) {
+  if (!this._emitter) this._emitter = new EventEmitter();
+  return this._emitter.once("finish", listener);
+};
+
+function getParentSpan(span: Span) {
+  if (span.parentSpanId) {
+    return span.spanRecorder?.spans?.find(
+      (s) => s.spanId === span.parentSpanId
+    );
+  }
+  return null;
+}
+
+/**
+ * Start a tracing span. The returned span should be explicitly ended with
+ * `finishSpan`. The created span will be a child of whatever span is currently
+ * active (and then become the current span itself), or if there is no current
+ * span or transaction, this will start one for you.
+ *
+ * Alternatively, can explicitly pass an actual span object to be the parent:
+ * `startSpan({ parentSpan: yourSpan })`. In this case, the new span won't
+ * automatically become the global "current" span.
+ *
+ * Spans created this way will be automatically canceled when their transaction
+ * finishes (Sentry will drop any unfinished spans). Alternatively, you can
+ * set `timeout` to a number of milliseconds, and the span will be canceled
+ * after that time (this should prevent you from accidentally leaving a span
+ * open forever). Canceled spans have a non-ok status set and a `cancel` tag
+ * with a reason.
+ *
+ * More on why spans need canceling:
+ * https://github.com/getsentry/sentry-javascript/issues/4165#issuecomment-971424754
+ */
+export function startSpan(options: SpanOptions): Span {
+  let { parentSpan, timeout, ...spanOptions } = options;
+
+  let scope;
+  if (!parentSpan) {
+    scope = Sentry.getCurrentHub().getScope();
+    parentSpan = scope.getSpan();
+  }
+
+  let newSpan: Span;
+  if (parentSpan) {
+    newSpan = parentSpan.startChild(spanOptions);
+  } else {
+    newSpan = Sentry.startTransaction(spanOptions as any);
+  }
+
+  // If we retrieved the span from the scope, update the scope.
+  if (scope) {
+    scope.setSpan(newSpan);
+  }
+
+  if (timeout) {
+    setTimeout(() => {
+      cancelSpan(newSpan, "deadline_exceeded");
+    }, timeout).unref();
+  } else if (newSpan.transaction !== newSpan) {
+    (newSpan.transaction as PatchedTransaction)?._onFinish(() =>
+      cancelSpan(newSpan, "cancelled", "did_not_finish")
+    );
+  }
+
+  return newSpan;
+}
+
+/**
+ * Finish a tracing span. This will also replace the global "current" span with
+ * this span's parent (if it has a parent).
+ */
+export function finishSpan(span: Span, timestamp?: number): void {
+  if (span.endTimestamp) return;
+
+  span.finish(timestamp);
+
+  const scope = Sentry.getCurrentHub().getScope();
+  if (scope.getSpan() === span) {
+    const parent = getParentSpan(span);
+    scope.setSpan(parent);
```

Comment:
In the case where the span has no parent, what happens? Is it okay to set to e.g. a null span?

Reply:
Yep — as long as nothing has gone *horribly* awry through misuse of the SDK somewhere, the only kind of span that doesn’t have a parent should be a transaction. When you end the transaction, you should remove it from the scope, which is what winds up happening here.